perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -974,3 +975,225 @@ def MoveFile(self, target, source_path, remote_path=''):
+  https://jujucharms.com/docs/stable/config-manual
+  """
+
+  OS_TYPE = os_types.JUJU
+
+  isController = False
+
+  """
+  A reference to the juju controller, useful when operations occur against
+  a unit's VM but need to be preformed from the controller.
+  """
+  controller = None
+
+  vm_group = None
+
+  machines = {}
+  units = []
+
+  installation_lock = threading.Lock()
+
+  environments_yaml = """
+  default: perfkit
+
+  environments:
+      perfkit:
+          type: manual
+          bootstrap-host: {0}
+  """
+
+  def _Bootstrap(self):
+    """Bootstrap a Juju environment"""
+    resp, _ = self.RemoteHostCommand("juju bootstrap")
+
+  def JujuAddMachine(self, unit):
+    """
+    Adds a manually-created virtual machine to Juju
+
+    Args:
+      unit: An object representing the unit's BaseVirtualMachine
+    """
+    resp, _ = self.RemoteHostCommand("juju add-machine ssh:%s" %
+                                     unit.internal_ip)
+
+    # We don't know what the machine's going to be used for yet,
+    # but track it's placement for easier access later.
+    # We're looking for the output: created machine %d
+    machine_id = _[_.rindex(' '):].strip()
+    self.machines[machine_id] = unit
+
+  def JujuConfigureEnvironment(self):
+    """Configure a bootstrapped Juju environment"""
+    if self.isController:
+      resp, _ = self.RemoteHostCommand("mkdir -p ~/.juju")
+
+      with vm_util.NamedTemporaryFile() as tf:
+        tf.write(self.environments_yaml.format(self.internal_ip))
+        tf.close()
+        self.PushFile(tf.name, "~/.juju/environments.yaml")
+
+  def JujuEnvironment(self):
+    """Get the name of the current environment"""
+    output, _ = self.RemoteHostCommand('juju switch')
+    return output.strip()
+
+  def JujuRun(self, cmd):
+    """Run a command on the virtual machine
+
+    Args:
+      cmd: The command to run
+    """
+    output, _ = self.RemoteHostCommand(cmd)
+    return output.strip()
+
+  def JujuStatus(self, pattern=''):
+    """Return the status of the Juju environment.
+
+    Args:
+      pattern: Optionally match machines/services with a pattern
+    """
+    output, _ = self.RemoteHostCommand('juju status %s --format=json' %
+                                       pattern)
+    return output.strip()
+
+  def JujuVersion(self):
+    """Return the Juju version"""
+    output, _ = self.RemoteHostCommand('juju version')
+    return output.strip()
+
+  def JujuSet(self, service, params=[]):
+    """Set the configuration options on a deployed service.
+
+    Args:
+      service: The name of the service
+      params: A list of key=values pairs
+    """
+    output, _ = self.RemoteHostCommand(
+        'juju set %s %s' % (service, ' '.join(params)))
+    return output.strip()
+
+  @vm_util.Retry(poll_interval=30, timeout=3600)
+  def JujuWait(self):
+    """
+    Waits for all deployed services to be installed, configured, and idle.
+
+    Args:
```

Comment:
nit: looks like you forgot to remove the argument documentation after you removed them